Projectiles

Pr0methean opened the pull request:
Also implements arrow consumption for ItemBow, updates the entity IDs for projectiles, and stubs out and registers missing projectile classes.

aramperes committed:
GlowProjectile, GlowEgg and GlowSnowball entities, ItemProjectile, fix ItemTimedUsage, handle UseItemMessage

aramperes committed:
Bow and Arrow, move velocity code to GlowLivingEntity

aramperes committed:
Make projectiles have an offset from the source (prevents self-colision)

aramperes committed:
Fix arrows and eggs (and bug eggs because why not)

Pr0methean committed:
First-pass partial implementation of sendMap

Will send blank maps until GlowMapRenderer.render() is implemented.

Pr0methean committed:
Fix a stray import

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into dev

Pr0methean committed:
Remove unused imports

Pr0methean committed:
Implement GlowMapRenderer.render() (#1)

This implementation is basic -- it only considers the material value of the highest block at each pixel -- but it implements the limited vision range and is configurable in YAML. Someone may still need to implement persistence for the GlowMapCanvas.

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/origin/dev' into newProjectiles

Pr0methean committed:
Stub out missing projectile entities

Also renames GlowEntity.getDirection() to getCardinalFacing() to resolve
a name collision

Pr0methean committed:
Bugfix: infinity bows do consume tipped/spectral arrows

Pr0methean committed:
Style fixes

Pr0methean committed:
Style fixes

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean commented:
This is a partial fix of #505

Pr0methean committed:
Revert stray changes to GlowMapRenderer

Pr0methean committed:
Implement most on-impact effects of projectiles

Also object IDs.

hibo98 commented on src/main/java/net/glowstone/entity/EntityRegistry.java:
```diff
@@ -68,74 +68,18 @@
-import org.bukkit.entity.Zombie;
-import org.bukkit.entity.ZombieHorse;
-import org.bukkit.entity.ZombieVillager;
+import org.bukkit.entity.*;
```
Please remove the wildcard import
See [Code Style](https://github.com/GlowstoneMC/Glowstone/wiki/Code-Style)

Pr0methean committed:
Fireball refinements

Pr0methean committed:
Arrow bug fixes

Dispenser arrows only deal 1 damage, and they can be picked up

Pr0methean committed:
EntityRegistry style fixes

mastercoms commented:
FYI, we'll be waiting to merge this after January 1st, due to the current soft feature freeze.

Pr0methean commented:
Thanks for the warning. I'll check again for merge conflicts shortly after Christmas.

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Copy PotionMeta data to TippedArrow

Copies Color, PotionData and List<PotionEffect>.

Pr0methean committed:
Copy more lingering-potion data to AreaEffectCloud

Copies Color & PotionData

Pr0methean committed:
Style fixes

Pr0methean commented:
Should an AreaEffectCloud created by a lingering potion always be the same color as the potion?

Pr0methean committed:
Merge branch 'dev' into newProjectiles

Pr0methean committed:
Lombokify GlowAreaEffectCloud

Pr0methean committed:
Lombokify projectiles

Pr0methean committed:
Implement GlowAreaEffectCloud.createSpawnMessage

Pr0methean committed:
Implement arrow enchantments, damage calculations, flame

All bow enchantments except knockback should now work.

Pr0methean committed:
Bows take damage when used

Also implements the unbreaking enchantment for bows and tools.

Pr0methean committed:
Style fixes

Pr0methean committed:
Adds AreaEffectCloud metadata messages

Pr0methean committed:
Fix unused imports

Pr0methean committed:
Implement lingering potions

Pr0methean committed:
Bugfix for multi-effect lingering potions

Pr0methean committed:
Style fix

Pr0methean committed:
Bug fix

Pr0methean committed:
Burning arrows can ignite TNT

Pr0methean committed:
Formatting fixes

Pr0methean committed:
Revert irrelevant changes to ArmorConstants

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

# Conflicts:

#	src/main/java/net/glowstone/dispenser/FlintAndSteelDispenseBehavior.java
#	src/main/java/net/glowstone/entity/EntityRegistry.java

#	src/main/java/net/glowstone/net/message/play/entity/SpawnObjectMessage.java

Pr0methean committed:
Merge fix

Pr0methean committed:
Fix bad merge

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Call GlowExperienceOrb.setFromBottle(true)

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

# Conflicts:

#	src/main/java/net/glowstone/net/handler/play/player/UseItemHandler.java

Pr0methean committed:
Fix bad merge of UseItemHandler

Pr0methean committed:
Factor out repeated expressions

Pr0methean committed:
Fix an NPE bug

Pr0methean committed:
Add ItemBowTest

Pr0methean committed:
Refactor: arrow is spent at END of firing

Pr0methean committed:
Remove unused import

Pr0methean committed:
Disable drag for fireballs

Pr0methean committed:
Merge branch 'dev' into newProjectiles

Pr0methean committed:
Merge the recent Ender Pearl work

Pr0methean committed:
Dragon fireballs have ID 93

Pr0methean committed:
Replace a few pulsePhysics overrides with setGravityAccel

Pr0methean committed:
Dispense behaviors for projectiles

Pr0methean committed:
Refactor ProjectileDispenseBehavior: don't really need subclass

Pr0methean committed:
Unwrap lines

Pr0methean committed:
Style fixes

Pr0methean committed:
More style fixes

Pr0methean commented:
This is ready for review.

Pr0methean committed:
Add missing projectile types

aramperes commented on src/main/java/net/glowstone/block/entity/state/GlowDispenser.java:
```diff
@@ -92,12 +98,24 @@ public static void register() {
                     InventoryUtil.copyPotionDataToTippedArrow(tippedArrow, itemStack);
                     return tippedArrow;
                 })));
-        registry.putBehavior(Material.FIREWORK_CHARGE, new ProjectileDispenseBehavior(location -> {
```
Was this removal intentional?

Pr0methean replied:
Yes: should have been FIREBALL.

aramperes replied:
Do fireworks no longer get let off by dispensers? Or is `firework_charge` something else?

Pr0methean replied:
I'm pretty sure FIREWORK_CHARGE is a firework *star*. The dispense behavior for firework *rockets* will not be a ProjectileDispenseBehavior either, because GlowFirework doesn't implement Projectile.

aramperes replied:
Ah okay, I thought the firework entities were considered projectiles.

Pr0methean replied:
Adding a TODO in the next commit.

aramperes commented on src/main/java/net/glowstone/block/blocktype/BlockTnt.java:
```diff
@@ -19,7 +20,7 @@
      * @param ignitedByExplosion True if another explosion caused this ignition.
      */
-    public static void igniteBlock(GlowBlock tntBlock, boolean ignitedByExplosion) {
+    public static void igniteBlock(Block tntBlock, boolean ignitedByExplosion) {
```
I think this is due to a merge mistake -- is there a reason for this change?

Pr0methean replied:
The block doesn't actually have to be a GlowBlock for the method body to work, and widening the types will help simplify calling code and maximize plugin compatibility.

aramperes commented on src/main/java/net/glowstone/block/itemtype/ItemBow.java:
```diff
@@ -0,0 +1,101 @@
+        player.setUsageTime(0);
+        currentArrow = null;
+    }
+
```
Remove whitespace.

Pr0methean replied:
Done in the next commit.

aramperes commented on src/main/java/net/glowstone/entity/GlowEntity.java:
```diff
@@ -207,6 +208,11 @@
+     * Whether friction applies to the entity.
+     */
+    @Setter
+    private boolean friction = true;
```
Why only lombokify the Setter?

Pr0methean replied:
Because a lombokified Getter would be called "isFriction" where we'd want "hasFriction". And the getter isn't used anyway, so for now there's none.

aramperes replied:
Ah okay :)

aramperes commented on src/main/java/net/glowstone/entity/projectile/GlowArrow.java:
```diff
@@ -0,0 +1,133 @@
+
+    public GlowArrow(Location location) {
+        super(location);
+        setGravityAccel(new Vector(0,-0.05,0));
```
Spaces after commas (`0, -0.05, 0`)

Pr0methean replied:
Done in the next commit.

aramperes commented on src/main/java/net/glowstone/entity/projectile/GlowArrow.java:
```diff
@@ -0,0 +1,133 @@
+            setVelocity(new Vector(0, 0, 0));
+        }
+    }
+
```
Remove whitespace.

Pr0methean replied:
Done in the next commit.

aramperes commented on src/main/java/net/glowstone/entity/projectile/GlowArrow.java:
```diff
@@ -0,0 +1,133 @@
+
+    @Override
+    public void setPickupRule(PickupRule pickupRule) {
+
```
Can you implement this? It shouldn't be too hard to just add a field with lombok.

Pr0methean replied:
Switching to interface defaults instead (stub generation must have been using a different version of Glowkit).

aramperes commented on src/main/java/net/glowstone/entity/projectile/GlowArrow.java:
```diff
@@ -0,0 +1,133 @@
+        metadata.setBit(MetadataIndex.ARROW_CRITICAL, 0x1, critical);
+    }
+
+    @Override public PickupStatus getPickupStatus() {
```
Annotations on different line

Pr0methean replied:
Done in the next commit.

aramperes commented on src/main/java/net/glowstone/entity/projectile/GlowArrow.java:
```diff
@@ -0,0 +1,133 @@
+                PickupStatus.ALLOWED;
+    }
+
+    @Override public void setPickupStatus(PickupStatus pickupStatus) {
```
Annotations on different line

Pr0methean replied:
Done in the next commit.

aramperes commented on src/main/java/net/glowstone/entity/projectile/GlowLingeringPotion.java:
```diff
@@ -0,0 +1,55 @@
+        }
+        remove();
+    }
+
```
Remove whitespace.

Pr0methean replied:
Done in the next commit.

aramperes commented on src/main/java/net/glowstone/io/entity/AreaEffectCloudStore.java:
```diff
@@ -0,0 +1,30 @@
+    @Override
+    public void load(GlowAreaEffectCloud entity, CompoundTag tag) {
+        super.load(entity, tag);
+        // TODO
```
Can you add what NBT fields need to be added in the TODO?

Pr0methean replied:
Done in the next commit.

aramperes commented on src/test/java/net/glowstone/block/itemtype/ItemBowTest.java:
```diff
@@ -0,0 +1,67 @@
+        // FIXME: assertEquals(1, arrows.getAmount()); // one arrow should be consumed
+    }
+
+}
```
Add blank line at end (LF)

Pr0methean replied:
Done in the next commit.

aramperes commented on src/test/java/net/glowstone/net/PlayProtocolTest.java:
```diff
@@ -84,6 +82,9 @@
 import org.bukkit.util.Vector;
 
+import java.util.Arrays;
+import java.util.UUID;
```
Probably due to a merge mistake

Pr0methean replied:
Reverted in the next commit.

aramperes commented on src/main/java/net/glowstone/block/itemtype/ItemBow.java:
```diff
@@ -0,0 +1,101 @@
+            switch (itemStack.getType()) {
+                case ARROW:
+                    if (currentArrow == null) {
+                        currentArrow = itemStack;
```
I don't think this is how it should be implemented. The arrow lookup should be done in the `endUse`: the startUse should just check if there *is* an arrow somewhere.

Suppose that the player has two stacks of arrows in their inventory. This function will determine that the first stack is to be used, and will be stored in `currentArrow`. If the stack is dropped or removed during the bow pulling (by a plugin or whatever), the `endUse` function won't be able to find the stack with this current system. This search should be done when by the `endUse`.

The `endUse` function should find the arrow and get the slot index for the stack. Then, it will summon the arrow and remove one/delete and update the slot (`inventory.setItem(slot, arrowStack)`)

Pr0methean replied:
Done in cb38717fadd2b1bdbaa3f0108b47cff6680c9c9d

Pr0methean committed:
Address Momo's review comments

Pr0methean committed:
Implement the double arrow check for ItemBow

aramperes commented on src/main/java/net/glowstone/block/itemtype/ItemBow.java:
```diff
@@ -0,0 +1,115 @@
+    public void endUse(GlowPlayer player, ItemStack item) {
+        ItemStack currentArrow = null;
+        // Check arrows again, since plugins can change the inventory while a bow is drawn
+        findArrow: for (ItemStack itemStack : player.getInventory().getContents()) {
```
Merge both for-loops into a method, returning an `Optional<ItemStack>`.

Pr0methean replied:
Done in 4a56ae374ea310c61123396fc796d71a0692da34

Pr0methean committed:
Extract findArrow method

Pr0methean committed:
Style fix

aramperes commented on src/main/java/net/glowstone/block/itemtype/ItemProjectile.java:
```diff
@@ -0,0 +1,20 @@
+        this.entityType = entityType;
+    }
+
+    public abstract Projectile use(GlowPlayer player, ItemStack holding);
```
Needs documentation

Pr0methean replied:
Documented in a314f6a8d42113df74766f2043f9dbc9bb23a91d

aramperes commented on src/main/java/net/glowstone/util/InventoryUtil.java:
```diff
@@ -67,4 +76,46 @@ public static int getRandomSlot(Random random, Inventory inventory, boolean igno
+     * @param player The player holding the item, or null if held by a dispenser
+     * @param holding The item
+     */
+    public static void damageItem(GlowPlayer player, ItemStack holding) {
```
What I would do, is make a method with just the `ItemStack`, that returns an `ItemStack`. The ItemStack doesn't get updated in the player's inventory until you use `inventory.setItem(slot, stack)`; and since the slot information isn't inside of the ItemStack object, this current implementation won't work.

Also, in that case, return `InventoryUtil.createEmptyStack()` if the item is to be removed (max damage)

Pr0methean replied:
Done in 87f6098857730d7a152df87c2d8ed49c15535b69

aramperes commented:
Also, please consider stubbing/implementing the `WitherSkull` entity, which extends `Fireball`

Pr0methean committed:
Partial response to review

Pr0methean committed:
DamageItem conversion

Pr0methean committed:
GlowWitherSkull

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Add missing ProjectileStore instances

Pr0methean commented:
Added GlowWitherSkull in f394fab

aramperes commented on src/main/java/net/glowstone/entity/projectile/GlowWitherSkull.java:
```diff
@@ -0,0 +1,22 @@
+        return NETWORK_ID;
+    }
+
+    public GlowWitherSkull(Location location) {
```
Put the constructor above other methods.

Pr0methean replied:
Done in f97a894

Pr0methean committed:
Move constructor to top

Pr0methean committed:
Remove unused import

Pr0methean committed:
Fix inventory checks in ItemBowTest

Pr0methean committed:
Improve ItemBowTest

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Eliminate an if-else

Pr0methean committed:
Extract PotionLike interface
Unifies PotionMeta->GlowMetaPotion and PotionMeta->GlowTippedArrow copying

Pr0methean committed:
GlowAreaEffectCloud implements PotionLike

Pr0methean committed:
Rename PotionLike to PotionBased

Pr0methean committed:
Move InventoryUtil.copyPotionData to PotionBased

Pr0methean committed:
Rename PotionBased to PotionDataHolder, move to root

Pr0methean committed:
Change copyPotionData to instance copyFrom

Pr0methean committed:
Fix AreaEffectCloudStore.createEntity

Pr0methean committed:
Style fixes

aramperes commented on src/main/java/net/glowstone/block/itemtype/ItemBow.java:
```diff
@@ -0,0 +1,116 @@
+
+    @Override
+    public void startUse(GlowPlayer player, ItemStack item) {
+        if (findArrow(player).isPresent()) {
```
In vanilla, you can shoot a bow even if you don't have arrows in your inventory. By default, it will shoot plain arrows, and if there are tipped arrows, it will use those instead (without consumption).

Here, the startUse would do nothing if the player is in creative and has no arrows, thus not shooting anything in the end.

Pr0methean replied:
Fixed: now shoots a plain arrow if you're in Creative and have no arrows.

Pr0methean committed:
In Creative, allow shooting without holding arrows

Pr0methean committed:
More tests

Pr0methean committed:
Add note to FIXME comment

Pr0methean committed:
Fix test bug

Pr0methean committed:
Uncomment parts of the test

Pr0methean committed:
Add test for Infinity enchantment

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
ItemExperienceBottle

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Postremus commented on src/main/java/net/glowstone/entity/projectile/GlowLingeringPotion.java:
```diff
@@ -0,0 +1,54 @@
+                // Cloud effects have only 1/4 the usual duration
+                PotionEffectType type = effect.getType();
+                if (!type.isInstant()) {
+                    effect = new PotionEffect(effect.getType(), effect.getDuration() >> 2,
```
Could we use a simple "/ 4" here instead?
Reading bit shifts is not not natural to all of us, and requires a few extra seconds to understand.

Pr0methean replied:
Done in bd87257c067ff07dde11c181fff35b4a7ce36355

Pr0methean committed:
Change ">> 2" to "/ 4"

Pr0methean committed:
Tests for ItemProjectile subclasses

Also ItemTypeTest, an abstract superclass for tests of other item types

Pr0methean committed:
Remove redundant arrow-speed code

Arrow velocity and criticality are set inside ItemBow.endUse()

Pr0methean committed:
Unused import

Pr0methean committed:
Missing javadocs

Pr0methean committed:
Add TODO for entity collisions

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Fix projectile-block collisions

Pr0methean committed:
Implement reapplication delay for AOE cloud

Pr0methean committed:
Refactor: switch to getNearbyEntities

Pr0methean committed:
Add comment

mastercoms commented on src/main/java/net/glowstone/entity/projectile/GlowProjectile.java:
```diff
@@ -0,0 +1,76 @@
+import net.glowstone.util.Position;
+import org.bukkit.Location;
+import org.bukkit.block.Block;
+import org.bukkit.block.BlockFace;
```
Unused import.

Pr0methean committed:
bind(new AreaEffectCloudStore()); [c

Pr0methean committed:
Unused import

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Remove magic number

Pr0methean committed:
Unify ender-pearl physics

Pr0methean committed:
Fix CheckStyle

Pr0methean committed:
Entity collisions

Pr0methean committed:
Don't suppress super.pulsePhysics if collide(LivingEntity) is no-op

Pr0methean committed:
Merge duplicate constant

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

# Conflicts:
#	src/main/java/net/glowstone/entity/GlowLivingEntity.java
#	src/main/java/net/glowstone/entity/objects/GlowEnderPearl.java

Pr0methean committed:
Fix bad merge

Pr0methean committed:
Update reference to GlowEntity.id -> entityId

mastercoms commented on src/main/java/net/glowstone/PotionDataHolder.java:
```diff
@@ -0,0 +1,126 @@
+    List<PotionEffect> getCustomEffects();
+
+    /**
+     * Adds a custom potion effect to this arrow.
```
Why do these docs refer to arrows only?

Pr0methean replied:
Fixed.

mastercoms commented on src/main/java/net/glowstone/block/ItemTable.java:
```diff
@@ -391,7 +395,6 @@ private void registerBuiltins() {
                 .YELLOW_GLAZED_TERRACOTTA));
         reg(Material.CHORUS_FLOWER, new BlockChorusFlower());
         reg(Material.CHORUS_PLANT, new BlockChorusPlant());
-
```
Don't remove the separation here between blocks and items.

Pr0methean replied:
Reverted and added comments.

mastercoms commented on src/main/java/net/glowstone/entity/GlowEntity.java:
```diff
@@ -957,28 +969,50 @@ protected void pulsePhysics() {
+                    } else {
+                        horizontalAirDrag = 0.91;
+                        velocity.setX(velocity.getX() * horizontalAirDrag);
+                        velocity.setZ(velocity.getZ() * 0.91);
```
Use `horizontalAirDrag` here as well.

Pr0methean replied:
Done. Also removed spurious setter on line 988.

mastercoms commented on src/main/java/net/glowstone/entity/GlowEntity.java:
```diff
@@ -957,28 +969,50 @@ protected void pulsePhysics() {
+                        velocity.setY(velocity.getY() + getGravityAccel().getY() / 4d);
+                        break;
+                    default:
+                        velocity.setY(velocity.getY() + getGravityAccel().getY() / 4d);
```
Gravity should not be divided by 4 here.

Pr0methean replied:
That's the existing behavior (see existing lines 964 and 967 below what's now 973).

mastercoms commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -663,11 +668,60 @@ public boolean isWithinSolidBlock() {
-        return entity;
+    public <T extends Projectile> T launchProjectile(Class<? extends T> clazz, Vector vector) {
+        float offset = 0.0F;
+        float velocity = 1.5F;
```
Rename velocity to speed here, and any where else velocity is being defined as a magnitude. Then rename vector to velocity.

Pr0methean replied:
Done in all throwProjectile overloads.

mastercoms commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -663,11 +668,60 @@ public boolean isWithinSolidBlock() {
+        return projectile;
+    }
+
+    protected <T extends Projectile> T throwProjectile(Class<? extends T> type, Location location,
```
Why does the method name change from launchProjectile to throwProjectile?

Likewise, why does clazz get changed to type?

Pr0methean replied:
Renamed to `launchProjectile` and `type` throughout.

mastercoms commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -663,11 +668,60 @@ public boolean isWithinSolidBlock() {
+
+    protected <T extends Projectile> T throwProjectile(Class<? extends T> type, Location location,
+            Vector originalVector, float offset, float velocity) {
+        double k = Math.toRadians(-1);
```
Could you rename these k variables to something more descriptive?

aramperes replied:
I made this variable to avoid using `Math.toRadians` for both pitch and yaw. In retrospect, there's a lot of stuff that can be simplified with this math (considering `cos(-x) = cos(x)`). I would definitely clean this up.

Pr0methean replied:
Done: Calling this one `negDegreesPerRadian` and the other one `magnitude`.

mastercoms commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -663,11 +668,60 @@ public boolean isWithinSolidBlock() {
+    protected <T extends Projectile> T throwProjectile(Class<? extends T> type, Location location,
+            Vector originalVector, float offset, float velocity) {
+        double k = Math.toRadians(-1);
+        double x = cos(k * location.getPitch()) * sin(k * location.getYaw());
```
Could you explain your vector math here and in the next method?

aramperes replied:
I believe I wrote this math originally in the projectiles branch. Here's my explanation (I learned trigonometry and vector math in another language, sorry if I'm not using the terms properly).

**Pitch/Yaw**
Pitch is defined as the angle of rotation on the horizontal axis, ranging from 90 degrees to -90 degrees.
Yaw is defined as the angle of rotation on the vertical axis ranging from 180 to -180.

**X component**
The first part consists of using the pitch (horizontal-axis rotation angle) to find out at which horizontal angle (downwards/upwards) the player is shooting the projectile. This angle helps as a “multiplier” for the X component, because shooting directly upwards or downwards should nullify the component. Because pitch is ranged negatively (negative is below the origin), we have to negate the conversion:
![image](https://user-images.githubusercontent.com/6775216/35294011-7bcaf83e-0043-11e8-9cbe-71768cf891b3.png)
We must then take into consideration the vertical-axis rotation angle (yaw). A yaw of -90deg represents a full X component (+X), while 90deg represents the opposite (-X). 180 and 0 degrees represent a null X component. This conversion can be achieved by using:
![image](https://user-images.githubusercontent.com/6775216/35294068-a761a628-0043-11e8-9a10-dff0a8b05f5f.png)
Because the pitch acts as a multiplier to the vertical-axis rotation:
![image](https://user-images.githubusercontent.com/6775216/35294084-b6209084-0043-11e8-8218-8d66a5e019cb.png)

**Z component**
This conversion is similar to the X component: it uses the pitch as a multiplier too. What changes is the use of the yaw range, which can be calculated using sin instead of cosine.
![image](https://user-images.githubusercontent.com/6775216/35294149-e250339e-0043-11e8-9896-e3864bd60663.png)

**Y component**
This is similar to the operation used to find the multiplier to the X/Z components, but here we need the upwards and downwards directions to represent a full Y component (using sin) instead of nullifying it (using cosine).
![image](https://user-images.githubusercontent.com/6775216/35294228-01fec70a-0044-11e8-98d0-eb1b4ad22ce3.png)

Pr0methean replied:
Added a comment to explain this briefly.

mastercoms commented on src/main/java/net/glowstone/entity/projectile/GlowArrow.java:
```diff
@@ -0,0 +1,130 @@
+            BlockTnt.igniteBlock(block, false);
+        }
+        setOnGround(true);
+        // Don't despawn
```
Arrows should be despawned eventually, perhaps with a configurable value like in Spigot/Paper.

Pr0methean replied:
Added TODO for this, while I find out the best way to schedule the despawn event.

mastercoms commented on src/main/java/net/glowstone/entity/projectile/GlowArrow.java:
```diff
@@ -0,0 +1,130 @@
+
+    @Override
+    protected int getObjectId() {
+        return 60;
```
Get rid of this object ID stuff and define it here: https://github.com/GlowstoneMC/Glowstone/blob/97fc5a6017b566af4e86b30cf6def6ebc4f0b9d4/src/main/java/net/glowstone/net/message/play/entity/SpawnObjectMessage.java#L14

Pr0methean replied:
Done for the return values of all overrides of getObjectId()

mastercoms commented on src/main/java/net/glowstone/entity/projectile/GlowEnderPearl.java:
```diff
@@ -0,0 +1,83 @@
+package net.glowstone.entity.projectile;
+
+import static net.glowstone.net.message.play.entity.SpawnObjectMessage.THROWN_ENDERPEARL;
```
I would prefer using `SpawnObjectMessage.THROWN_ENDERPEARL` in line.

Pr0methean replied:
Done.

mastercoms commented on src/main/java/net/glowstone/entity/projectile/GlowEnderPearl.java:
```diff
@@ -0,0 +1,83 @@
+import org.bukkit.util.Vector;
+
+public class GlowEnderPearl extends GlowProjectile implements EnderPearl {
+    private static final double ENDER_PEARL_DAMAGE = 5.0;
```
Add a line between the class declaration and this field.

Pr0methean replied:
Done.

mastercoms commented on src/main/java/net/glowstone/entity/meta/MetadataIndex.java:
```diff
@@ -80,16 +82,15 @@
-    //AREAEFFECTCLOUD_PARTICLE_PARAM2(11, INT, Entity.class),
+    AREAEFFECTCLOUD_RADIUS(6, FLOAT, AreaEffectCloud.class),
+    AREAEFFECTCLOUD_COLOR(7, INT, AreaEffectCloud.class),
+    AREAEFFECTCLOUD_UNKNOWN(8, BOOLEAN, AreaEffectCloud.class),
```
Rename this to `AREAEFFECTCLOUD_POINT`.

Pr0methean replied:
Done.

mastercoms commented on src/main/java/net/glowstone/net/handler/play/player/UseItemHandler.java:
```diff
@@ -90,6 +92,16 @@ static void handleRightClickAir(GlowPlayer player, ItemStack holding, EquipmentS
                 if (type.getContext().isAirApplicable()) {
                     type.rightClickAir(player, holding);
+                } else {
+                    if ((holding.getType() == Material.WATER_BUCKET
```
Seems like this was incorrectly reverted from the ender pearl PR.

Pr0methean replied:
Fixed.

mastercoms commented on src/main/java/net/glowstone/scheduler/GlowScheduler.java:
```diff
@@ -74,17 +76,33 @@
      *
      * @param server The server that will use this scheduler.
-     * @param worlds The {@link WorldScheduler} this scheduler will use for ticking the server's
+     * @param worlds The {@link WorldScheduler} this scheduler will use for ticking the servers
```
The previous doc was correct: server's worlds

Pr0methean replied:
Fixed on both constructors.

mastercoms commented on src/test/java/net/glowstone/block/itemtype/ItemProjectileTest.java:
```diff
@@ -0,0 +1,42 @@
+    protected void checkProjectile(T projectile) {
+        verify(player, times(1)).launchProjectile(projectileClass);
+    }
+}
```
Newline

Pr0methean replied:
Done.

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -663,11 +668,60 @@ public boolean isWithinSolidBlock() {
+
+    private <T extends Projectile> T throwProjectile(Class<? extends T> clazz, Location location,
+            double x, double y, double z, float velocity) {
+        double k = Math.sqrt(x * x + y * y + z * z);
```
`k` here is the magnitude of the unit represented by `x, y, z`, previously calculated. I believe the math below multiplies the vector by the given `velocity` magnitude (using component operations) and then adds these results to the base components, in order to give a certain offset. (this is in response to mastercoms)

Pr0methean replied:
Now called `magnitude` and checked for zero (so if x, y and z are all zero, we don't end up at coordinates of infinity or NaN).

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Add TODO re arrow despawn

Pr0methean committed:
Move network IDs to SpawnObjectMessage

Pr0methean committed:
Address review comments

aramperes commented on src/main/java/net/glowstone/entity/GlowLivingEntity.java:
```diff
@@ -654,20 +659,101 @@ public boolean isWithinSolidBlock() {
+     */
+    protected <T extends Projectile> T launchProjectile(Class<? extends T> type, Location location,
+            Vector originalVector, float pitchOffset, float velocity) {
+        double negDegreesPerRadian = Math.toRadians(-1);
```
I think these calculations can be simplified like this:

```java
double pitchRadians = Math.toRadians(location.getPitch());
double yawRadians = Math.toRadians(location.getYaw());

double verticalMultiplier = cos(pitchRadians);
double x = verticalMultiplier * sin(-yawRadians);
double z = verticalMultiplier * cos(yawRadians);
double y = sin(-(Math.toRadians(location.getPitch() - pitchOffset));
```

kaenganxt commented on src/main/java/net/glowstone/net/handler/play/player/UseItemHandler.java:
```diff
@@ -90,6 +92,10 @@ static void handleRightClickAir(GlowPlayer player, ItemStack holding, EquipmentS
                 if (type.getContext().isAirApplicable()) {
                     type.rightClickAir(player, holding);
+                } else if (type instanceof ItemTimedUsage) {
+                    ((ItemTimedUsage) type).startUse(player, holding);
```
This call is already in the `rightClickAir` handler of `ItemTimedUsage`. You should make ItemProjectile similar to ItemTimedUsage (returning `Context.ANY` for `getContext()` and calling the `use` method in both `rightClickAir` and `rightClickBlock`). I haven't tested it, but I think currently you are not able to shoot a projectile when aiming at a block.

Pr0methean replied:
Done in 985f213b34b6752c1058d5d2f37d824a9eacaed3

Pr0methean committed:
Arrows on ground eventually despawn

Also stub out effect on redstone inputs  buttons, pressure plates,
tripwires)

Pr0methean committed:
Refactor

Pr0methean committed:
Arrow storage

Pr0methean committed:
Bugfix: no longer has type parameter

Pr0methean committed:
Style fix

Pr0methean committed:
ItemProjectile/UseItemHandler refactors

Per
https://github.com/GlowstoneMC/Glowstone/pull/614#discussion_r163382496

Pr0methean committed:
Refactor launchProjectile

Per
https://github.com/GlowstoneMC/Glowstone/pull/614#discussion_r163380927

Pr0methean committed:
Syntax error

Pr0methean committed:
Unused imports

Pr0methean committed:
consumeItem refactors

Pr0methean committed:
Bugfix: arrow is not held when firing bow

Pr0methean committed:
FireballStore

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Store fireball yield/incendiary in custom NBT

Don't seem to exist in Mojang NBT

Pr0methean committed:
"bukkit" prefix convention

Pr0methean committed:
Style fix

Pr0methean committed:
Revert ""bukkit" prefix convention"

This reverts commit ed39210b389f384374135afdb107a534e575972c.

Pr0methean committed:
Bukkit has ExplosionPower(int)

Pr0methean committed:
Default yield 1

Pr0methean committed:
SplashPotionStore

Pr0methean committed:
Widen consumeHeldItem param type

Pr0methean committed:
Remove unnecessary qualifier

Pr0methean committed:
Unused import

Pr0methean committed:
Add PotionEffectHolder interface

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

Pr0methean committed:
Revert "Add PotionEffectHolder interface"

This reverts commit df7950896a3136029f8fafeee69dd8cb317ea52a.

Pr0methean committed:
Implement AreaEffectCloudStore loader

Pr0methean committed:
Finish NormalTippedArrowStore

Pr0methean committed:
Syntax error

Pr0methean committed:
Style fixes

Pr0methean committed:
Improve warning message

Pr0methean committed:
Rename read*IfPresent to handle*IfPresent

Pr0methean committed:
Refactor

Pr0methean committed:
Add FloatConsumer interface to prevent boxing

Pr0methean committed:
Javadoc bug fix

Pr0methean committed:
Style fix

aramperes commented on src/main/java/net/glowstone/EventFactory.java:
```diff
@@ -49,15 +52,16 @@ private EventFactory() {
     public static <T extends Event> T callEvent(T event) {
-        GlowServer server = (GlowServer) Bukkit.getServer();
+        Server server = Bukkit.getServer();
 
```
Why was this change necessary? `Bukkit.getServer()` is guaranteed to be a `GlowServer` instance. The only difference I see here are the additional imports.

Or was this done because it broke tests somehow?

Pr0methean replied:
Yes it did (ServerShim doesn't extend GlowServer).

mastercoms commented on src/main/java/net/glowstone/io/entity/AreaEffectCloudStore.java:
```diff
@@ -52,8 +52,8 @@ public void load(GlowAreaEffectCloud entity, CompoundTag tag) {
             } catch (IllegalArgumentException e) {
-                Bukkit.getServer().getLogger().warning("Ignoring invalid particle type "
-                        + particle);
+                Bukkit.getServer().getLogger().warning(String.format(
```
Could you use `.info(Level.WARNING, ...)`

mastercoms commented on src/main/java/net/glowstone/io/entity/AreaEffectCloudStore.java:
```diff
@@ -0,0 +1,111 @@
+            try {
+                entity.setParticle(Particle.valueOf(particle));
+            } catch (IllegalArgumentException e) {
+                Bukkit.getServer().getLogger().warning(String.format(
```
Could you use `log(Level.WARNING, ...)`

Pr0methean replied:
Done in aaadd0e2356f861f5a5090c2d63ee61f501c23f2.

mastercoms commented on src/main/java/net/glowstone/io/entity/FireballStore.java:
```diff
@@ -0,0 +1,45 @@
+
+public class FireballStore<T extends GlowFireball> extends ProjectileStore<T> {
+
+    public static final String IS_INCENDIARY = "X-Glowstone-IsIncendiary";
```
Could you make these private like the other ones? Also, I feel like `glowstone:IsIncendiary` or if that's not valid, `glowstoneIsIncendiary` would be better.

Pr0methean replied:
Made private and changed to "glowstone:IsIncendiary" and "glowstone:ExplosionPowerFloat" in aaadd0e2356f861f5a5090c2d63ee61f501c23f2.

Pr0methean committed:
Tweaks per latest review

Pr0methean committed:
Refactor logging from AreaEffectCloudStore

Per
https://github.com/GlowstoneMC/Glowstone/commit/aaadd0e2356f861f5a5090c2d63ee61f501c23f2#r27171608

Pr0methean committed:
Merge remote-tracking branch 'refs/remotes/GlowstoneMC/dev' into newProjectiles

mastercoms commented on src/main/java/net/glowstone/block/itemtype/ItemBow.java:
```diff
@@ -0,0 +1,127 @@
+                // Not in creative mode and have no arrow
+                break;
+            default:
+                player.getServer().getLogger()
```
`GlowServer.logger` (or is this for testing) and `log(Level, Supplier)`

Pr0methean replied:
Done in f656da22b66794f0e340518025296d49d53a43a8 (also migrated to the `Supplier<String>` logging function).

Pr0methean committed:
Refactor ItemBow to use Supplier<String> logging function

mastercoms commented on src/main/java/net/glowstone/io/entity/AreaEffectCloudStore.java:
```diff
@@ -0,0 +1,113 @@
+package net.glowstone.io.entity;
+
+import java.util.UUID;
+import java.util.logging.Level;
```
unused import

Pr0methean replied:
Fixed in b599a16b37602fe0c22e8a8ca6a53d49bb3cff91

mastercoms commented on src/main/java/net/glowstone/io/entity/AreaEffectCloudStore.java:
```diff
@@ -0,0 +1,113 @@
+import net.glowstone.entity.GlowAreaEffectCloud;
+import net.glowstone.inventory.GlowMetaPotion;
+import net.glowstone.util.nbt.CompoundTag;
+import org.bukkit.Bukkit;
```
unused import

Pr0methean replied:
Fixed in b599a16b37602fe0c22e8a8ca6a53d49bb3cff91

Pr0methean committed:
Unused imports
